device/alpaka/src/clusterization/clusterization_algorithm.cpp

```diff
@@ -0,0 +1,136 @@
+        auto const blockExtent =
+            ::alpaka::getWorkDiv<::alpaka::Block, ::alpaka::Threads>(acc)[0u];
+
+        auto& partition_start =
+            ::alpaka::declareSharedVar<unsigned int, __COUNTER__>(acc);
+        auto& partition_end =
+            ::alpaka::declareSharedVar<unsigned int, __COUNTER__>(acc);
+        auto& outi = ::alpaka::declareSharedVar<unsigned int, __COUNTER__>(acc);
+
+        device::details::index_t* const shared_v =
+            ::alpaka::getDynSharedMem<device::details::index_t>(acc);
+        vecmem::data::vector_view<device::details::index_t> f_view{
+            max_cells_per_partition, shared_v};
+        vecmem::data::vector_view<device::details::index_t> gf_view{
+            max_cells_per_partition, shared_v + max_cells_per_partition};
+
+        alpaka::barrier<TAcc> barry_r(&acc);
+
+        device::ccl_kernel(localThreadIdx, blockExtent, localBlockIdx,
+                           cells_view, modules_view, max_cells_per_partition,
+                           target_cells_per_partition, partition_start,
+                           partition_end, outi, f_view, gf_view, barry_r,
+                           measurements_view, cell_links);
+    }
+};
+
+clusterization_algorithm::clusterization_algorithm(
+    const traccc::memory_resource& mr, vecmem::copy& copy,
+    const unsigned short target_cells_per_partition)
+    : m_target_cells_per_partition(target_cells_per_partition),
+      m_mr(mr),
+      m_copy(copy) {}
+
+clusterization_algorithm::output_type clusterization_algorithm::operator()(
+    const cell_collection_types::const_view& cells,
+    const cell_module_collection_types::const_view& modules) const {
+
+    // Setup alpaka
+    auto devHost = ::alpaka::getDevByIdx(::alpaka::Platform<Host>{}, 0u);
+    auto devAcc = ::alpaka::getDevByIdx(::alpaka::Platform<Acc>{}, 0u);
+    auto queue = Queue{devAcc};
+
+    // Number of cells
+    const cell_collection_types::view::size_type num_cells =
+        m_copy.get().get_size(cells);
+
+    // Create the result object, overestimating the number of measurements.
+    measurement_collection_types::buffer measurements{
+        num_cells, m_mr.main, vecmem::data::buffer_type::resizable};
+    m_copy.get().setup(measurements)->ignore();
+
+    // If there are no cells, return right away.
+    if (num_cells == 0) {
+        return measurements;
+    }
+
+    // Create buffer for linking cells to their measurements.
+    //
+    // @todo Construct cell clusters on demand in a member function for
+    // debugging.
+    //
+    vecmem::data::vector_buffer<unsigned int> cell_links(num_cells, m_mr.main);
+    m_copy.get().setup(cell_links)->ignore();
+
+    // Launch ccl kernel. Each thread will handle a single cell.
+    const device::details::ccl_kernel_helper helper{
+        m_target_cells_per_partition, num_cells};
+    auto workDiv =
+        makeWorkDiv<Acc>(helper.num_partitions, helper.threads_per_partition);
```

Comment:
This will also need some updating with #595. I recognise that you have been working on this for a _long_ time, though, so I don't blame you if you want to get this in first.